src/IK/src/QPInverseKinematics.cpp

```diff
@@ -0,0 +1,417 @@
+            m_pimpl->tasks.erase(taskName);
+
+            return false;
+        }
+
+        // add the weight
+        m_pimpl->tasks[taskName].weight = weight.value();
+
+        // add the task to the list of the element that are used to build the cost function
+        m_pimpl->costs.push_back(m_pimpl->tasks[taskName]);
+    } else
+    {
+        m_pimpl->constraints.push_back(m_pimpl->tasks[taskName]);
+    }
+
+    return true;
+}
+
+bool QPInverseKinematics::finalize(const System::VariablesHandler& handler)
+{
+    constexpr auto logPrefix = "[QPInverseKinematics::finalize]";
+
+    // set the variable handler for all the tasks
+    std::size_t numberOfConstraints = 0;
+    for (auto& [name, task] : m_pimpl->tasks)
+    {
+        if (!task.task->setVariableHandler(handler))
+        {
+            log()->error("{} Error while setting the variable handler in the task named {}, having "
+                         "the following description {}.",
+                         logPrefix,
+                         name,
+                         task.task->getDescription());
+            return false;
+        }
+
+        // if priority is equal to 0 the task is considered as hard constraint.
+        if (task.priority == 0)
+        {
+            numberOfConstraints += task.task->size();
+        }
+    }
+
+    // resize the temporary matrix usefull to reduce dynamics allocation when advance() is called
+    for (auto& cost : m_pimpl->costs)
+    {
+        cost.get().tmp.resize(handler.getNumberOfVariables(), cost.get().weight.size());
+    }
+
+    m_pimpl->solver.data()->setNumberOfVariables(handler.getNumberOfVariables());
+    m_pimpl->solver.data()->setNumberOfConstraints(numberOfConstraints);
+
+    // resize matrices
+    m_pimpl->hessian.resize(handler.getNumberOfVariables(), handler.getNumberOfVariables());
+    m_pimpl->gradient.resize(handler.getNumberOfVariables());
+
+    m_pimpl->constraintMatrix.resize(numberOfConstraints, handler.getNumberOfVariables());
+    m_pimpl->upperBound.resize(numberOfConstraints);
+    m_pimpl->lowerBound.resize(numberOfConstraints);
+
+    if (!handler.getVariable(m_pimpl->robotVelocityVariable.name, m_pimpl->robotVelocityVariable))
+    {
+        log()->error("{} Error while retrieving the robot velocity variable.", logPrefix);
+        return false;
+    }
+
+    return true;
+}
+
+std::vector<std::string> QPInverseKinematics::getTaskNames() const
+{
+    std::vector<std::string> tasksName;
+
+    for (const auto& [name, task] : m_pimpl->tasks)
+    {
+        tasksName.push_back(name);
+    }
+
+    return tasksName;
+}
+
+bool QPInverseKinematics::advance()
+{
+    constexpr auto logPrefix = "[QPInverseKinematics::advance]";
+
+    // when advance is called the previous solution is no more valid
+    m_pimpl->isValid = false;
+
+    if (!m_pimpl->isInitialized)
+    {
+        log()->error("{} Please call initialize() before advance().", logPrefix);
+        return false;
+    }
```

Comment:
I think that it is worth checking that ``finalize`` has been called too.

Reply:
Done here a8329da9